File: rust/Problem10.rs
/*
 * The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.
 *
 * Find the sum of all the primes below two million.
 */

fn sieve(is_composites: &mut [bool]) {
    is_composites[0] = true;
    is_composites[1] = true;
    let sqrt_limit = (is_composites.len() as f64).sqrt() as usize;
    for i in 2..sqrt_limit {
        if !is_composites[i] {
            for j in i..(((is_composites.len() as f64) / i as f64).ceil() as usize) {
                is_composites[j * i] = true;
            }
        }
    }
}

fn main() {
    let mut is_composites = [false; 2000000];
    sieve(&mut is_composites);
    let answer: usize = is_composites.iter()
        .enumerate()
        .filter_map(|(i, &is_composite)| if is_composite { None } else { Some(i) })
        .sum();
    println!("{}", answer);
}

File: java/Problem10.java
/*
 * The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.
 *
 * Find the sum of all the primes below two million.
 */
import java.util.PrimitiveIterator;
import java.util.Spliterator;
import java.util.Spliterators;
import java.util.stream.IntStream;
import java.util.stream.StreamSupport;

public class Problem10 {
    private static class Primes implements Iterable<Integer> {
        private final int limit;
        private boolean[] isComposites;

        public Primes(int limit) {
            this.limit = limit;
        }

        @Override
        public PrimitiveIterator.OfInt iterator() {
            sieveIfNeeded();
            return new PrimesIterator();
        }

        @Override
        public Spliterator.OfInt spliterator() {
            return Spliterators.spliterator(iterator(),
                    (long) Math.floor(limit / Math.log(limit)),
                    Spliterator.DISTINCT | Spliterator.IMMUTABLE | Spliterator.NONNULL | Spliterator.ORDERED | Spliterator.SIZED | Spliterator.SORTED);
        }

        private synchronized void sieveIfNeeded() {
            if (isComposites == null) {
                isComposites = new boolean[limit];
                isComposites[0] = true;
                isComposites[1] = true;
                int sqrtLimit = (int)Math.sqrt(limit);
                for (int i = 2; i <= sqrtLimit; i++) {
                    if (!isComposites[i]) {
                        for (int j = i * i; j < limit; j += i) {
                            isComposites[j] = true;
                        }
                    }
                }
            }
        }

        private class PrimesIterator implements PrimitiveIterator.OfInt {
            private static final int NOT_FOUND = Integer.MIN_VALUE;
            private int next = 2;

            @Override
            public boolean hasNext() {
                return next != NOT_FOUND;
            }

            @Override
            public int nextInt() {
                int current = next;
                for (int i = current + 1; i < isComposites.length; i++) {
                    if (!isComposites[i]) {
                        next = i;
                        return current;
                    }
                }
                next = NOT_FOUND;
                return current;
            }
        }
    }

    public static void main(String[] args) {
        Primes primes = new Primes(2000000);
        IntStream stream = StreamSupport.intStream(primes.spliterator(), false);
        long answer = stream.mapToLong(Long::new).sum();
        System.out.println(answer);
    }
}
